composer/trainer/trainer.py

```diff
@@ -956,8 +959,10 @@ def __init__(
         if eval_dataloader is None:
             evaluators: List[Evaluator] = []
         else:
+            model_metric_names = [str(k) for k in model.metrics(train=False).keys()]  # type: ignore
             evaluators = [
-                ensure_evaluator(evaluator, model.metrics(train=False)) for evaluator in ensure_tuple(eval_dataloader)
+                ensure_evaluator(evaluator, default_metric_names=model_metric_names)
```

Comment:
I'm a bit confused by the logic here -- if a user has different metrics they want to apply to different evaluators, how does that work? Seems like here we just run every metric against every eval dataset?

Reply:
The `ensure_evaluator()` function actually does two things: if the given `evaluator` is an `Evaluator` instance (meaning it has its own set of metrics), the `evaluator` is just returned. Otherwise, if the given `evaluator` is just a dataloader, it is converted to an `Evaluator` that evaluates against all the given `default_metric_names` (which by default are just set to be all the evaluation metrics defined in the model).